Blog post: Advanced Flink Application Patterns Vol.2 (#310)
apache/flink-web: _posts/2020-02-26-demo-fraud-detection-2.md

---
layout: post
title: "Advanced Flink Application Patterns Vol.2:

> Case Study of a Fraud Detection System"

Author: Yes, I agree. Changing it to:
"Advanced Flink Application Patterns Vol.2: Dynamic Updates of Application Logic". I could also modify it accordingly for the first one, but I am not sure if it makes much sense.

date: 2020-02-26T12:00:00.000Z
authors:
- alex:
  name: "Alexander Fedulov"
  twitter: "alex_fedulov"
categories: news
excerpt: In this series of blog posts you will learn about powerful Flink patterns for building streaming applications.
---

In the first article of the series **[!LINK]**, we gave a high-level description of the objectives and required functionality of a Fraud Detection engine. We described how Apache Flink's data-partitioning mechanism makes it possible to determine the distribution of data in a cluster not based on a hardcoded `KeysExtractor` implementation, but using a list of modifiable rules instead.

We intentionally omitted details of how the applied rules are initialized and what possibilities exist for updating them at runtime. In this post, we will address exactly these details. You will learn how the approach described in Part 1 **[!LINK]** can be applied in combination with a dynamic configuration that makes it possible to avoid recompiling the code with every modification of the business logic execution parameters.

> # Rules Broadcasting

Reviewer: I still miss a bit structure in terms of additional headings. At the moment, everything (except start and end) is under the "Rules Broadcasting" heading. Adding more here would make it simpler for the reader to know where (s)he is and what comes next or even skip sections with things (s)he already knows.

> # Rules Broadcasting

Author: Not true - I've moved it to the same level as the next one
## Data Exchange inside Apache Flink

> # Rules Broadcasting

Author: Damn, the last changes that I had in the working directory did not get commited - I've added another header to separate the "detour" section:

`## Data Exchange inside Apache Flink`

And moved this one to the same header level.

        .process(new DynamicAlertFunction())
```

`DynamicKeyFunction`, as described earlier, provides dynamic data partitioning, while `DynamicAlertFunction` is responsible for executing the main logic of processing transactions and sending alert messages according to defined rules.

For a simplified description, we previously assumed that the applied set of rules are pre-initialized and are accessible via the `List<Rules>` within DynamicKeyFunction.

```java
public class DynamicKeyFunction
    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {

  /* Simplified */
  List<Rule> rules = /* Rules that are initialized somehow.*/;
  ...
}
```

Obviously, adding rules to this list is possible directly inside the code of the Flink Job at the stage of its initialization (create a List object, use it's `add` method). A major drawback of doing so is that it will require recompilation of the job with each rule modification. In the real Fraud Detection system, you can expect to modify these rules on a frequent basis, which makes this approach unacceptable from the point of view of business and operational requirements - a different approach is needed.

> Let's next have a look at the example of DSL of one of the sample rules presented in the previous post:

Reviewer: How about the proposed heading here?

> Let's next have a look at the example of DSL of one of the sample rules presented in the previous post:

Author: I've added it locally, but it did not get "published", sorry. The changes are there now.

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rule-dsl.png" width="800px" alt="Figure 1: Rules DSL"/>
<br/>
<i><small>Figure 1: Rules DSL</small></i>
</center>
<br/>

> ![](assets/rule-dsl.png)

Author: I left if so that Marcos can also see the images in Github "preview". I will remove those before merging.

Earlier we have already seen how part of this definition, namely `groupingKeyNames`, is used in `DynamicKeyFunction` to extract message keys. At the same time, all other parameters of this rule are necessary within `DynamicAlertFunction` - they define the actual logic of performed operations and it's parameters (such as the alert triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use a special data distribution mechanism in Apache Flink - [events broadcast](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/state/broadcast_state.html). Let’s turn to the Job Graph of the system that we are building:

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph"/>
<br/>

> <i><small>Figure 2: Job Graph</small></i>

Author: "Figure 2: Job Graph of the Fraud Detection Flink Job"
? @NicoK

</center>
<br/>
![](assets/job-graph.png)

On it you can see the main blocks of our Transactions processing pipeline:<br>

> * **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>

Author: Hm.. 
```
The main blocks of our Transactions processing pipeline are:
 - Transaction Source consumes transaction messages from Kafka partitions in parallel
```
This sounds a big off. Maybe "Transaction Source, which consumes transaction messages ..." 
@NicoK

> * **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>

Author: I see now that I used ", consuming" to avoid having "which, which, which" all the bullet points.

> * **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>

Reviewer: yes - my two separate comments/suggestions were probably not in sync :/

> * **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>

Reviewer: I think, your suggestion of "Transaction Source, which consumes..." is better than the original. You didn't apply it yet though, did you?

> * **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>

Author: I wanted to avoid repetition in all three components "which, which, which". I've checked with a native speaker - sounds OK to him.

>     * The forward operator following it means that all data consumed by one of the parallel instances of Transaction Source operator will be transferred to exactly one instance of the subsequent DynamicKeyFunction operator, without redistribution. It also indicates the same level of parallelism of two connected operators (12 in this case).

Author: Yes, I was not not sure what to call it. I guess reformulating it according to your suggestion hides the fact that we do not have a noun for it :) 
I am, however, not sure about the task/operator terminology when we are talking about the `DynamicKeyFunction`. I was orienting myself on the new Glossary:

Task - 
https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#task

Operator - 
https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#operator

The way I understand it is that one "instance" of an Operator is a Task, so I would probably stick to my original version here.

@NicoK

>     * The forward operator following it means that all data consumed by one of the parallel instances of Transaction Source operator will be transferred to exactly one instance of the subsequent DynamicKeyFunction operator, without redistribution. It also indicates the same level of parallelism of two connected operators (12 in this case).

Reviewer: Your operators are not chained and thus deployed in separate tasks. Communication between tasks is via the network stack via network buffers (in case of forwards, these will thus be exchanged inside a single JVM) while communication between (chained) operators is done directly by calling the next operator with a copy or reference of the output from the current one.
You may see chaining as a special case of forwarding or a fourth communication pattern.
Either way, your job has some chained operators where you do not explain the message passing. The "blocks" you are talking about are tasks.

>     * The forward operator following it means that all data consumed by one of the parallel instances of Transaction Source operator will be transferred to exactly one instance of the subsequent DynamicKeyFunction operator, without redistribution. It also indicates the same level of parallelism of two connected operators (12 in this case).

Reviewer: In terms of operator vs. task; an operator is a node of the logical graph and a task is a node of the physical graph. So, technically, there are no communications between operators. You could say "operator instances" everywhere though but we still ignore one important communication pattern: the chain. If you explain communication patterns between operator instances, I think, this one should be in there. Still, I'd rather see tasks and not go into chaining for this blog post (or maybe remove the "normal" communication patterns for good and maybe describe them in the docs instead if there is nothing yet)

 <center>
 <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/forward.png" width="800px" alt="Figure 3: Message passing across operators : FORWARD"/>
 <br/>

>  <i><small>Figure 3: Message passing across operators : FORWARD</small></i>

Author: Same Glossary question.

 </center>
 <br/>
![](assets/forward.png)

> * **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator performs `hashing` of this dynamic key and corresponding data sharding

Author: Hmm... https://www.grammarly.com/blog/comma-before-which/.

> * **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator performs `hashing` of this dynamic key and corresponding data sharding

Reviewer: Hard to say what the "sentence" is here and in enumerations of that style with sub-bullets. I guess, it would even work without the "which", but it all depends how you write this block in general. I won't fight over a comma ;)

> * **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator performs `hashing` of this dynamic key and corresponding data sharding

Author: My understanding is that a "task" is already one of the parallel instances and it cannot be further broken down, so "among all parallel instances of the following task" is not an applicable description.

> * **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator performs `hashing` of this dynamic key and corresponding data sharding

Reviewer: Regarding the comma, if you add one before "which", you may also want to add one before "that" which is essentially the same.

    * hash here means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator.

 <center>

>  <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across operators : HASH (keyBy)"/>

Author: We definitely need to sync on the task/operator terminology ;)
I use the same visual encoding as in the first part of the blog - circles are transactions (https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html)
On this image dotted lines are tasks ("Tasks encapsulate exactly one parallel instance of an Operator or Operator Chain."). Maybe I should also add links to the Glossary when I first mention "task" and "operator".

>  <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across operators : HASH (keyBy)"/>

Reviewer: Sure, you could add links if you want to. In the end, when reading this post, I think, readers would, however take whatever term you present as fore-granted. Nevertheless it should be the correct one ;)
Think about the simpler image but you could also keep (and adapt) the current one which conveys the message but takes some more space. If you keep the current one, you may also need to think about adding the remaining arrows because otherwise the image is incomplete.

>  <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across operators : HASH (keyBy)"/>

Author: @NicoK which arrows do you mean? Currently each event (orange circle) is dispatched according to it's hash code - I do not see which additional errors could be necessary.

>  <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across operators : HASH (keyBy)"/>

Reviewer: I actually mixed up what the circles meant. If it is a transaction/event, it is ok to keep the arrows as they are (I thought they indicate all communication paths and are missing some).
In that case, Operator_1_2's event may miss its hash. Does it make sense to draw dotted lines for the paths which are not shown? Or is that cluttering the image too much?

The proposal you did not talk about though is whether it makes sense to simplify the image by having Operator_1 with p=1 only and instead two events in there.

>  <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across operators : HASH (keyBy)"/>

Author: I have added a sentence with explicit explanation, so that people also do not get confused.
So, regarding the hash - I actually gave it a try when I was drawing it. It did not fit on the right because of all the arrows and I decide not to bother. If you think it unclear like that, I can move the hash "bubble" to the left. 
Regarding p=1 
- I though about about it, but, as you said, I need two circles within the same task - this makes it stand out from all the other figures, I did not like that idea. Additionally, I feel that this variant looses part of the message that this process happens in parallel in every task.
- Then I though about having p = 2, this also makes things awkward: we either forward both circles to the same receiving end, which leaves one just hanging there, or forward them to one respectively, either making them look like a forward, or a bow tie, which somehow also makes it easy to miss the point.
This is how I came to this variant. I think it is the best option. I am open to adding the missing "hash" if you think it is important.

>  <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across operators : HASH (keyBy)"/>

Reviewer: no, the "hash" is not too important; also thanks for the explanation regarding p=1 - makes sense

 <br/>

>  <i><small>Figure 4: Message passing across operators : HASH (keyBy)</small></i>

Author: --||--

 </center>
 <br/>
![](assets/hash.png)

* **DynamicAlertFunction** that accumulates a data window and creates Alerts based on it
    * Rebalance means that the data redistribution is caused not by the explicit `keyBy` operator, but by a change in parallelism (12 -> 1).

<center>
 <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rebalance.png" width="800px" alt="Figure 4: Message passing across operators : REBALANCE"/>
 <br/>

>  <i><small>Figure 5: Message passing across operators : REBALANCE</small></i>

Author: --||--

```java
public class DynamicKeyFunction
    extends BroadcastProcessFunction<Transaction, Rule, Keyed<Transaction, String, Integer>> {

  @Override
  public void processBroadcastElement(Rule rule,
                                     Context ctx,
                                     Collector<Keyed<Transaction, String, Integer>> out) {
    BroadcastState<Integer, Rule> broadcastState = ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
    broadcastState.put(rule.getRuleId(), rule);
  }

  @Override
  public void processElement(Transaction event,
                           ReadOnlyContext ctx,
                           Collector<Keyed<Transaction, String, Integer>> out){
    ReadOnlyBroadcastState<Integer, Rule> rulesState =
                                  ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
    for (Map.Entry<Integer, Rule> entry : rulesState.immutableEntries()) {
        final Rule rule = entry.getValue();
        out.collect(
          new Keyed<>(
            event, KeysExtractor.getKey(rule.getGroupingKeyNames(), event), rule.getRuleId()));
    }
  }
}
```

> In the above code `processElement()` receives Transactions and `processBroadcastElement()` receives Rules updates. When a new rule is created, it is distributed and saved in all parallel instances of the operator using `processBroadcastState`, as depicted in Figure 6. We use Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, iteration is performed over entries in the dynamically updated broadcast state.

Reviewer: ```suggestion
In the code above, `processElement()` receives Transactions and `processBroadcastElement()` receives Rules updates. When a new rule is created, it is distributed and saved in all parallel instances of the operator using `processBroadcastState`, as depicted in Figure 6. We use a Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, iteration is performed over entries in the dynamically updated broadcast state.
```
One thing that is still a bit misleading is that "Figure 6" does not show `processBroadcastState`...

Author: " `processBroadcastState`, as depicted in Figure 6." ->
" `processBroadcastState`, in such a manner as depicted in Figure 6." ?

Reviewer: no, maybe "it is distributed as in Figure 6 and saved in..."
(I was a bit confused by the placement - Fig. 6 shows the distribution, not the saving part)

Author: https://forum.wordreference.com/threads/see-the-above-below-picture-or-see-the-picture-above-below.1301337/

> `DynamicAlertFunction` follows the same logic with respect to storing the rules in the broadcast MapState. As described in Part 1  **[!LINK]**, each message in the `processElement` input is intended to be processed by one specific rule and comes "pre-marked" with a corresponding ID by  `DynamicKeyFunction`. All we need to do is retrieve the definition of the corresponding rule from `BroadcastState` by the provided ID and process it according to the logic required by that rule. At this stage, we will also add messages to the internal function state in order to perform calculations on the required time window of data. We will consider how it is done in the final blog of the series about Fraud Detection.

Reviewer: add Link!
```suggestion
`DynamicAlertFunction` follows the same logic with respect to storing the rules in the broadcast `MapState`. As described in Part 1  **[!LINK]**, each message in the `processElement` input is intended to be processed by one specific rule and comes "pre-marked" with a corresponding ID created by `DynamicKeyFunction`. All we need to do is retrieve the definition of the corresponding rule from `BroadcastState` and process it accordingly. At this stage, we will also add messages to the internal function state in order to perform calculations on the required time window of data. This will be described in detail in the third and final part of this series on our Fraud Detection case study.
```

# Summary

> In this blogpost, we have continued our investigation of the use case of the Fraud Detection System built with Apache Flink. We have looked into different ways in which data can be distributed between Flink operators, and specifically into details of events broadcast. We have demonstrated, how `DynamicDataPartitioning` pattern described in the first part of the series **[LINK]** can be combined and enhanced by the functionality provided by the broadcast state pattern. The ability to send dynamic updates at runtime is a powerful feature of Apache Flink, which is applicable in a variety of other use cases, such as:

Reviewer: add link, and...
```suggestion
In this blogpost, we continued our investigation of the use case of a Fraud Detection System built with Apache Flink. We looked into different ways in which data can be distributed between tasks and, most importantly, examined broadcast state. We demonstrated how dynamic partitioning—a pattern described in the first part of the series **[LINK]**—can be combined and enhanced by the functionality provided by the broadcast state pattern. The ability to send dynamic updates at runtime is a powerful feature of Apache Flink, which is applicable in a variety of other use cases, such as:
```

- Control of state (cleanup/insert/fix)
- A/B experiments

> - Updates of ML model coefficients

Reviewer: This actually currently does not render a list and I'm not sure a bulleted list would be best in a summary. Maybe integrate this into the text?

Author: @MarkSfik , are there any style guidelines regarding this?

Reviewer 2: There are no specific guidelines here, to be honest. 

I agree it might be a bit strange to end the story with a bullet list. 

Maybe an alternative would be adding this as part of the paragraph, so finishing the paragraph like: 

"… which is applicable in a variety of other use cases, such as (a) Control of state (cleanup/insert/fix) (b) running A/B experiments or (c) executing updates of ML model coefficients."

I can make another suggestion if this is more appropriate. 
Thanks

</center>
<br/>

![](./../img/blog/patterns-blog-2/rule-dsl.png)
The previous post covered the use of `groupingKeyNames` by `DynamicKeyFunction` to extract message keys. Parameters from the second part of this rule are used by `DynamicAlertFunction`: they define the actual logic of the performed operations and their parameters (such as the alert triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use the [broadcast data distribution mechanism](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/stream/state/broadcast_state.html) of Apache Flink. Let’s turn to the Job Graph of the system that we are building:
<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph of the Fraud Detection Flink Job"/>
<br/>
<i><small>Figure 2: Job Graph of the Fraud Detection Flink Job</small></i>
</center>
<br/>

![](./../img/blog/patterns-blog-2/job-graph.png)

The main blocks of the Transactions processing pipeline are:<br>

* **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>

> * **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator hashes this dynamic key and partitions the data accordingly among all parallel instances of the following task

Reviewer: ```suggestion
* **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` hashes this dynamic key and partitions the data accordingly among all parallel instances of the following task
```

* **Dynamic Alert Function** that accumulates a data window and creates Alerts based on it

The job graph also depicts various data exchange patterns between the operators.

> * The __FORWARD__ connection following the Transaction Source means that all data consumed by one of the parallel instances of the Transaction Source operator is transferred to exactly one instance of the subsequent `DynamicKeyFunction` operator, without redistribution. It also indicates the same level of parallelism of the two connected operators (12 in this case). This communication pattern is depicted in Figure 3. Orange circles represent transactions and dotted rectangles depict parallel instances of the operators (also called [*tasks*](https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#operator)).  

Author: That, in contrast to hash there is no explicit per-message operation that defines its path. Do you have an idea how to make it more clear?

> * The __FORWARD__ connection following the Transaction Source means that all data consumed by one of the parallel instances of the Transaction Source operator is transferred to exactly one instance of the subsequent `DynamicKeyFunction` operator, without redistribution. It also indicates the same level of parallelism of the two connected operators (12 in this case). This communication pattern is depicted in Figure 3. Orange circles represent transactions and dotted rectangles depict parallel instances of the operators (also called [*tasks*](https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#operator)).  

Reviewer: Maybe just remove this? Because, although forward is typically a local communication, it doesn't have to be; so it may actually also be distributed over network to another slot/TM. You basically already described the 1:1 mapping in the sentence.

<center>

> <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/forward.png" width="800px" alt="Figure 3: Message passing across tasks: FORWARD"/>

Author: Reverted.

<br/>
<i><small>Figure 3: Message passing across tasks: FORWARD</small></i>
</center>
<br/>

![](./../img/blog/patterns-blog-2/forward.png)

> * The __HASH__ connection between the `DynamicKeyFunction` and the `DynamicAlertFunction` means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator. Such connection needs to be explicitly "requested" from Flink by using a `keyBy` operator.

Reviewer: ```suggestion
* The __HASH__ connection between the `DynamicKeyFunction` and the `DynamicAlertFunction` means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator. Such connection needs to be explicitly "requested" from Flink by using `keyBy`.
```

Author: Sometimes it gets awkward with names that end with a proper noun, like "Function", because when you read them, in your head it sounds like you have to slap an article there :)

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across tasks: HASH (via `keyBy`)"/>
<br/>
<i><small>Figure 4: Message passing across tasks: HASH (keyBy)</small></i>
</center>
<br/>

![](./../img/blog/patterns-blog-2/hash.png)

> * A __REBALANCE__ distribution is either caused by an explicit call to `rebalance()` or by a change of parallelism (12 -> 1 in case of the job graph from Figure 2). Calling `rebalance()` causes data to be repartitioned randomly and can help to mitigate data skew in certain scenarios.

Author: I guess .shuffle() also cause a REBALANCE, so probably good to mentioned it to, right?

> * A __REBALANCE__ distribution is either caused by an explicit call to `rebalance()` or by a change of parallelism (12 -> 1 in case of the job graph from Figure 2). Calling `rebalance()` causes data to be repartitioned randomly and can help to mitigate data skew in certain scenarios.

Reviewer: Actually, `keyBy()`, `partitionCustom()`, `rebalance()`, `rescale()`, and `shuffle()` are all different partitioning schemes which cause a network shuffle. Not sure which one(s) you want to describe here, but `rebalance()`, `rescale()`, and `shuffle()` are certainly in a group of non-key-based shuffling mechanisms.
* However, you may not want to put them under a same label or if so, then it shouldn't be called "rebalance" because this is already a specific scheme our of these three.
* Be sure to describe the right one (the current description is wrong).
* Maybe also start with the "what" before you describe the "how"?

> * A __REBALANCE__ distribution is either caused by an explicit call to `rebalance()` or by a change of parallelism (12 -> 1 in case of the job graph from Figure 2). Calling `rebalance()` causes data to be repartitioned randomly and can help to mitigate data skew in certain scenarios.

Reviewer: test comment: REBALANCE is being used when changing the parallelism (one may think, this leads to RESCALE)

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rebalance.png" width="800px" alt="Figure 4: Message passing across tasks: REBALANCE"/>
<br/>
<i><small>Figure 5: Message passing across tasks: REBALANCE</small></i>
</center>
<br/>

![](./../img/blog/patterns-blog-2/rebalance.png)

> At the top of the job graph in Figure 2 you can see an additional data source - **Rules Source**, which also consumes from Kafka. Rules are "mixed into" the main processing data flow through the `broadcast` channel. Unlike other methods of transmitting data between operators, such as `forward`, `hash` or `rebalance`, which make each message available for processing in only one of the parallel instances of the receiving operator, `broadcast` makes each message available at the input of all of the parallel instances of the operator which the _broadcast stream_ is connected to. This makes `broadcast` applicable to a wide range of tasks that need to affect the processing of all messages, regardless of their key or source partition.

Author: It is because there are a lot of images in the description - the section itself is 13 lines of text. I do not have a fresh idea neither how to restructure it nor how it explain it better. Sometimes people even have to flip the pages to see the relevant figure - seems unavoidable sometimes.

> At the top of the job graph in Figure 2 you can see an additional data source - **Rules Source**, which also consumes from Kafka. Rules are "mixed into" the main processing data flow through the `broadcast` channel. Unlike other methods of transmitting data between operators, such as `forward`, `hash` or `rebalance`, which make each message available for processing in only one of the parallel instances of the receiving operator, `broadcast` makes each message available at the input of all of the parallel instances of the operator which the _broadcast stream_ is connected to. This makes `broadcast` applicable to a wide range of tasks that need to affect the processing of all messages, regardless of their key or source partition.

Reviewer: Hmm...maybe you could re-use the jobgraph here and present the "Rules Source", "Transaction Source", and "Dynamic Partitioning" again as an example (just these three tasks and the arrows between them). Then, you could say that this is the same for the alert function and drop the code that creates this?

---
layout: post
title: "Advanced Flink Application Patterns Vol.2:
Dynamic Updates of Application Logic"
date: 2020-02-26T12:00:00.000Z
authors:
- alex:
  name: "Alexander Fedulov"
  twitter: "alex_fedulov"
categories: news
excerpt: In this series of blog posts you will learn about powerful Flink patterns for building streaming applications.
---

In the [first article](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html) of the series, we gave a high-level description of the objectives and required functionality of a Fraud Detection engine. We also described how to make data partitioning in Apache Flink customizable based on modifiable rules instead of using a hardcoded `KeysExtractor` implementation.

> We intentionally omitted details of how the applied rules are initialized and what possibilities exist for updating them at runtime. In this post, we will address exactly these details. You will learn how the approach to data partitioning described in [Part 1](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html) can be applied in combination with a dynamic configuration. These two patterns used together can eliminate the need to recompile the code and redeploy your Flink job for a wide range of modifications of the business logic.

Reviewer 2: ```suggestion
We intentionally omitted details of how the applied rules are initialized and what possibilities exist for updating them at runtime. In this post, we will address exactly these details. You will learn how the approach to data partitioning described in [Part 1](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html) can be applied in combination with a dynamic configuration. These two patterns, when used together, can eliminate the need to recompile the code and redeploy your Flink job for a wide range of modifications of the business logic.
```

## Rules Broadcasting

Let's first have a look at the [previously-defined](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html#dynamic-data-partitioning) data-processing pipeline:

```java
DataStream<Alert> alerts =
    transactions
        .process(new DynamicKeyFunction())
        .keyBy((keyed) -> keyed.getKey());
        .process(new DynamicAlertFunction())
```

`DynamicKeyFunction` provides dynamic data partitioning while `DynamicAlertFunction` is responsible for executing the main logic of processing transactions and sending alert messages according to defined rules.

> Vol.1 of this case study simplified the use case and assumed that the applied set of rules is pre-initialized and is accessible via the `List<Rules>` within `DynamicKeyFunction`.

Reviewer 2: ```suggestion
Vol.1 of this series simplified the use case and assumed that the applied set of rules is pre-initialized and accessible via the `List<Rules>` within `DynamicKeyFunction`.
```

```java
public class DynamicKeyFunction
    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {

  /* Simplified */
  List<Rule> rules = /* Rules that are initialized somehow.*/;
  ...
}
```

> Adding rules to this list is obviously possible directly inside the code of the Flink Job at the stage of its initialization (Create a `List` object; use it's `add` method). A major drawback of doing so is that it will require recompilation of the job with each rule modification. In a real Fraud Detection system, rules are expected to change on a frequent basis, and makes this approach unacceptable from the point of view of business and operational requirements. A different approach is needed.

Reviewer 2: ```suggestion
Adding rules to this list is obviously possible directly inside the code of the Flink Job at the stage of its initialization (Create a `List` object; use it's `add` method). A major drawback of doing so is that it will require the recompilation of the job with each rule modification. In a real Fraud Detection system, rules are expected to change on a frequent basis, making this approach unacceptable both from a business point-of-view and its operational requirements. A different approach is thus necessary.
```

Author: I would like to keep "business requirements" - it is something that has specific meaning when talking about software.

Reviewer 2: Thanks @afedulov 
I will remove the suggestion.

> Let's next have a look at a sample rule definition that we introduced in the previous post:

Reviewer 2: ```suggestion
Next, let's take a look at a sample rule definition that we introduced in the previous post of the series:
```

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rule-dsl.png" width="800px" alt="Figure 1: Rule definition"/>
<br/>
<i><small>Figure 1: Rule definition</small></i>
</center>
<br/>

![](./../img/blog/patterns-blog-2/rule-dsl.png)

> The previous post covered use of `groupingKeyNames` by `DynamicKeyFunction` to extract message keys. Parameters from the second part of this rule are used by `DynamicAlertFunction`: they define the actual logic of the performed operations and their parameters (such as the alert-triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use the [broadcast data distribution mechanism](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/stream/state/broadcast_state.html) of Apache Flink.

Reviewer 2: ```suggestion
The previous post covered the use of `groupingKeyNames` by `DynamicKeyFunction` to extract message keys. Parameters from the second part of this rule are used by `DynamicAlertFunction`: they define the actual logic of the performed operations and their parameters (such as the alert-triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use the [broadcast data distribution mechanism](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/stream/state/broadcast_state.html) of Apache Flink.
```

Author: @MarkSfik , I cannot spot the diff ...

Reviewer 2: correct, I will remove this comment. 
thanks @afedulov

Figure 2 presents the final job graph of the system that we are building:

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph of the Fraud Detection Flink Job"/>
<br/>
<i><small>Figure 2: Job Graph of the Fraud Detection Flink Job</small></i>
</center>
<br/>

![](./../img/blog/patterns-blog-2/job-graph.png)

The main blocks of the Transactions processing pipeline are:<br>

> * **Transaction Source**, consuming transaction messages from Kafka partitions in parallel. <br>

Reviewer 2: ```suggestion
* **Transaction Source** that consumes transaction messages from Kafka partitions in parallel. <br>
```

Author: I wanted to avoid using "X that ..." in all three bullet points. Is my version hard to understand?

Reviewer 2: this is more because it's not consistent, the bullet points have 

- **Transaction Source**, consuming
- **Dynamic Key Function** that performs data....
- **Dynamic Alert Function** that accumulates a....

> * **Dynamic Key Function** that performs data enrichment with a dynamic key. The subsequent `keyBy` hashes this dynamic key and partitions the data accordingly among all parallel instances of the following operator.

Reviewer 2: ```suggestion
* **Dynamic Key Function**, performs data enrichment with a dynamic key. The subsequent `keyBy` hashes this dynamic key and partitions the data accordingly among all parallel instances of the following operator.
```

* **Dynamic Alert Function** that accumulates a data window and creates Alerts based on it.

## Data Exchange inside Apache Flink

> The job graph above also indicates various data exchange patterns between the operators. In order to understand how the broadcast pattern works, let's take a short detour and discuss what methods of messages propagation exist in Apache Flink's distributed runtime.

Reviewer 2: ```suggestion
The job graph above also indicates various data exchange patterns between the operators. In order to understand how the broadcast pattern works, let's take a short detour and discuss what methods of message propagation exist in Apache Flink's distributed runtime.
```

```java
public class DynamicKeyFunction
    extends BroadcastProcessFunction<Transaction, Rule, Keyed<Transaction, String, Integer>> {

  @Override
  public void processBroadcastElement(Rule rule,
                                     Context ctx,
                                     Collector<Keyed<Transaction, String, Integer>> out) {
    BroadcastState<Integer, Rule> broadcastState = ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
    broadcastState.put(rule.getRuleId(), rule);
  }

  @Override
  public void processElement(Transaction event,
                           ReadOnlyContext ctx,
                           Collector<Keyed<Transaction, String, Integer>> out){
    ReadOnlyBroadcastState<Integer, Rule> rulesState =
                                  ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
    for (Map.Entry<Integer, Rule> entry : rulesState.immutableEntries()) {
        final Rule rule = entry.getValue();
        out.collect(
          new Keyed<>(
            event, KeysExtractor.getKey(rule.getGroupingKeyNames(), event), rule.getRuleId()));
    }
  }
}
```

> In the above code, `processElement()` receives Transactions, and `processBroadcastElement()` receives Rules updates. When a new rule is created, it is distributed as depicted in Figure 6 and saved in all parallel instances of the operator using `processBroadcastState`. We use a Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, we iterate over entries in the dynamically-updated broadcast state.

Reviewer 2: ```suggestion
In the above code, `processElement()` receives Transactions, and `processBroadcastElement()` receives Rule updates. When a new rule is created, it is distributed as depicted in Figure 6 and saved in all parallel instances of the operator using `processBroadcastState`. We use a Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, we iterate over entries in the dynamically-updated broadcast state.
```

`DynamicAlertFunction` follows the same logic with respect to storing the rules in the broadcast `MapState`. As described in [Part 1](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html), each message in the `processElement` input is intended to be processed by one specific rule and comes "pre-marked" with a corresponding ID by  `DynamicKeyFunction`. All we need to do is retrieve the definition of the corresponding rule from `BroadcastState` by using the provided ID and process it according to the logic required by that rule. At this stage, we will also add messages to the internal function state in order to perform calculations on the required time window of data. We will consider how this is done in the final blog of the series about Fraud Detection.

# Summary

> In this blogpost, we continued our investigation of the use case of a Fraud Detection System built with Apache Flink. We looked into different ways in which data can be distributed between parallel operator instances and, most importantly, examined broadcast state. We demonstrated how dynamic partitioning — a pattern described in the [first part](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html) of the series — can be combined and enhanced by the functionality provided by the broadcast state pattern. The ability to send dynamic updates at runtime is a powerful feature of Apache Flink that is applicable in a variety of other use cases, such as:

Reviewer 2: ```suggestion
In this blog post, we continued our investigation of the use case of a Fraud Detection System built with Apache Flink. We looked into different ways in which data can be distributed between parallel operator instances and, most importantly, examined broadcast state. We demonstrated how dynamic partitioning — a pattern described in the [first part](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html) of the series — can be combined and enhanced by the functionality provided by the broadcast state pattern. The ability to send dynamic updates at runtime is a powerful feature of Apache Flink that is applicable in a variety of other use cases, such as:
```

Author: Here just "blog post", right?

Reviewer 2: yes, that's correct @afedulov
